compaction_manager: stop_tasks, stop_ongoing_compactions: ignore errors

bhalevy opened the pull request:
stop() methods, like destructors must always succeed,
and returning errors from them is futile as there is
nothing else we can do with them by continue with shutdown.

stop_ongoing_compactions, in particular, currently returns the status
of stopped compaction tasks from `stop_tasks`, but still all tasks
must be stopped after it, even if they failed, so assert that
and ignore the errors.

Fixes scylladb/scylladb#21159

* Needs backport to 6.2 and 6.1, as commit 8cc99973ebaf9d664c4c2a20b8ce3f1b2eb7100a causes handles storage that might cause compaction tasks to fail and eventually terminate on shudown when the exceptions are thrown in noexcept context in the deferred stop destructor body

regevran commented on compaction/compaction_manager.cc:
```diff
@@ -1127,12 +1127,21 @@ future<> compaction_manager::drain() {
+    try {
+        do_stop();
+        if (auto cm = std::exchange(_task_manager_module, nullptr)) {
+            auto f = cm->stop();
```
```suggestion
            auto f = co_await cm->stop();
```
Before the change we await on `stop()` and we return whatever there is in `stop_future`.
After the change we await on `stop()` only if there was something in `stop_future` - otherwise we don't wait on `stop()`.

bhalevy replied:
> Before the change we await on stop() and we return whatever there is in stop_future.

The return type is `furure<>` so we don't really return anything (in _stop_future).

> After the change we await on stop() only if there was something in stop_future - otherwise we don't wait on stop().

why?
The change moves the future returned by cm->stop() into _stop_future and then it is co_await'ed as before.

regevran commented on compaction/compaction_manager.cc:
```diff
@@ -1127,12 +1127,21 @@ future<> compaction_manager::drain() {
+            co_return co_await std::move(*_stop_future);
+        }
+    } catch (...) {
+        cmlog.error("Failed to stop the manager: {}", std::current_exception());
```
"eliminating" all exceptions (with `...`) may be useful to prevent crashes if we assume that the system is in a stable enough state to continue.
Here we don't handle the exception, we simply report about it.
I don't know if there is something better to do, maybe just start report all these `...` catches with some grepable text for later investigation.

bhalevy replied:
we have nothing better to do other than reporting the failure.
we can restrict catch to particular errors but I'm not sure it's worth it

scylladb-promoter commented:
#### :red_circle: CI State: FAILURE

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/12796/)
:white_check_mark: - [Container Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/12796/)
:white_check_mark: - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/4457/)
:white_check_mark: - [dtest with tablets](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-tablets/1388/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/12479/)
:x: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/12796/)
#### Failed Tests (27/36394):
- [[empty]](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/12796/testReport/junit/boost.sstable_compaction_test.simple_backlog_controller_test_leveled.1.xunit/xml/Tests___Unit_Tests____empty_) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+[empty])
- [[empty]](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/12796/testReport/junit/boost.sstable_compaction_test.simple_backlog_controller_test_leveled.1.xunit/xml/Tests___Unit_Tests____empty__2) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+[empty])
- [[empty]](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/12796/testReport/junit/boost.sstable_compaction_test.simple_backlog_controller_test_leveled.1.xunit/xml/Tests___Unit_Tests____empty__3) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+[empty])
- [[empty]](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/12796/testReport/junit/boost.sstable_compaction_test.simple_backlog_controller_test_leveled.2.xunit/xml/Tests___Unit_Tests____empty_) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+[empty])
- [[empty]](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/12796/testReport/junit/boost.sstable_compaction_test.simple_backlog_controller_test_leveled.2.xunit/xml/Tests___Unit_Tests____empty__2) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+[empty])
- [[empty]](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/12796/testReport/junit/boost.sstable_compaction_test.simple_backlog_controller_test_leveled.2.xunit/xml/Tests___Unit_Tests____empty__3) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+[empty])
- [[empty]](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/12796/testReport/junit/boost.sstable_compaction_test.simple_backlog_controller_test_leveled.3.xunit/xml/Tests___Unit_Tests____empty_) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+[empty])
- [[empty]](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/12796/testReport/junit/boost.sstable_compaction_test.simple_backlog_controller_test_leveled.3.xunit/xml/Tests___Unit_Tests____empty__2) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+[empty])
- [[empty]](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/12796/testReport/junit/boost.sstable_compaction_test.simple_backlog_controller_test_leveled.3.xunit/xml/Tests___Unit_Tests____empty__3) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+[empty])
- [[empty]](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/12796/testReport/junit/boost.sstable_compaction_test.simple_backlog_controller_test_size_tiered.1.xunit/xml/Tests___Unit_Tests____empty_) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+[empty])
#### Build Details:
- Duration: 3 hr 58 min
- Builder: spider8.cloudius-systems.com

avikivity commented on compaction/compaction_manager.cc:
```diff
@@ -1127,12 +1127,21 @@ future<> compaction_manager::drain() {
-    if (_state != state::none) {
-        co_return co_await std::move(*_stop_future);
+    try {
+        do_stop();
```
do_stop() should prevent throwing on its own.

bhalevy replied:
it does.  I'm afraid the error came from the task manager stop() call, not from do_stop

avikivity replied:
Then keep do_stop() outside the try/catch, and fix task_manager::stop.

Deexie replied:
The error comes from `co_return co_await std::move(*_stop_future);`:

`do_stop` sets `_stop_future` to `really_do_stop`
`do_really_stop` co_awaits `stop_ongoing_compactions`

https://github.com/scylladb/scylladb/blob/88e2a7ea192b5963ee03f5ce4975d355e28c32c4/compaction/compaction_manager.cc#L1154

And `stop_ongoing_compactions` has noexcept trick:
https://github.com/scylladb/scylladb/blob/88e2a7ea192b5963ee03f5ce4975d355e28c32c4/compaction/compaction_manager.cc#L1112-L1114

elcallio replied:
Slightly off topic, but isn't a noexcept function returning an exception future somewhat breaking its contract (at least w.r. to a co_await caller)?

regevran replied:
> Slightly off topic, but isn't a noexcept function returning an exception future somewhat breaking its contract (at least w.r. to a co_await caller)?

Technically no, it _returns_ and not _throws_.
The contract here is that the `future` may contain many things, one of them is an exception.
So in this sense this is also not a contract violation.

bhalevy replied:
Agreed with @regevran.  There's no equivalent for `noexcept` for an async function that may return no errors.

bhalevy replied:
exceptions are evil

avikivity replied:
but evil isn't an exception

avikivity commented on compaction/compaction_manager.cc:
```diff
@@ -1127,12 +1127,21 @@ future<> compaction_manager::drain() {
+    try {
+        do_stop();
+        if (auto cm = std::exchange(_task_manager_module, nullptr)) {
+            auto f = cm->stop();
```
compaction_manager::stop() (and all stop() methods) should be responsible for their own error handling. Just like close() methods do. If we need to collect errors, there should be a flush() method called ahead.

bhalevy commented:
In v2 ([88e2a7e](https://github.com/scylladb/scylladb/commit/88e2a7ea192b5963ee03f5ce4975d355e28c32c4)):
- forced awaiting on task manager module stop before _stop_future
  - fixing ci failures in `sstable_compaction_test.simple_backlog_controller_test_{leveled,size_tiered}`
- simplified condition for awaiting _stop_future (we shouldn't care about the state, only if it's engaged or not)
- improved error message text
  - and demoted it to a warning

avikivity commented on compaction/compaction_manager.cc:
```diff
@@ -1127,12 +1127,23 @@ future<> compaction_manager::drain() {
+    } catch (...) {
+        cmlog.warn("Ignored error while stopping the manager: {}", std::current_exception());
     }
 }
```
If all we're waiting for here is task_manager and compaction_manager, then the error handling should be moved there.

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/12800/)
:white_check_mark: - [Container Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/12800/)
:white_check_mark: - [dtest with tablets](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-tablets/1390/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/12481/)
:white_check_mark: - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/4459/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/12800/)
#### Build Details:
- Duration: 3 hr 57 min
- Builder: spider5.cloudius-systems.com

bhalevy commented:
> Missing backport paperwork.

@avikivity, I wrote the following in the description.  Is it enough, or did you miss it by any chance?
> * Needs backport to 6.2 and 6.1, on top of [8cc9997](https://github.com/scylladb/scylladb/commit/8cc99973ebaf9d664c4c2a20b8ce3f1b2eb7100a)

scylladb-promoter commented:
#### :red_circle: CI State: FAILURE

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/12835/)
:white_check_mark: - [Docker Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci-docker/4/)
:x: - [dtest with tablets](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-tablets/1425/)
:x: - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/4495/)
:x: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/12518/)
:x: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/12835/)
#### Failed Tests (369/36532):
- [test_auth_v2_during_recovery.debug.1](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/12835/testReport/junit/%28root%29/test_auth_v2_migration/Tests___Unit_Tests___test_auth_v2_during_recovery_debug_1) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_auth_v2_during_recovery.debug.1)
- [test_coordinator_queue_management.debug.2](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/12835/testReport/junit/%28root%29/test_coordinator_queue_management/Tests___Unit_Tests___test_coordinator_queue_management_debug_2) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_coordinator_queue_management.debug.2)
- [test_feature_deprecation_works.debug.3](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/12835/testReport/junit/%28root%29/test_deprecating_cluster_features/Tests___Unit_Tests___test_feature_deprecation_works_debug_3) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_feature_deprecation_works.debug.3)
- [test_features_suppress_works.debug.1](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/12835/testReport/junit/%28root%29/test_deprecating_cluster_features/Tests___Unit_Tests___test_features_suppress_works_debug_1) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_features_suppress_works.debug.1)
- [test_features_suppress_works.debug.3](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/12835/testReport/junit/%28root%29/test_deprecating_cluster_features/Tests___Unit_Tests___test_features_suppress_works_debug_3) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_features_suppress_works.debug.3)
- [test_schema_versioning_with_recovery.dev.2](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/12835/testReport/junit/%28root%29/test_group0_schema_versioning/Tests___Unit_Tests___test_schema_versioning_with_recovery_dev_2) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_schema_versioning_with_recovery.dev.2)
- [test_maintenance_mode.debug.1](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/12835/testReport/junit/%28root%29/test_maintenance_mode/Tests___Unit_Tests___test_maintenance_mode_debug_1) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_maintenance_mode.debug.1)
- [test_maintenance_mode.debug.2](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/12835/testReport/junit/%28root%29/test_maintenance_mode/Tests___Unit_Tests___test_maintenance_mode_debug_2) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_maintenance_mode.debug.2)
- [test_mv_build_during_shutdown.debug.2](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/12835/testReport/junit/%28root%29/test_mv_fail_building/Tests___Unit_Tests___test_mv_build_during_shutdown_debug_2) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_mv_build_during_shutdown.debug.2)
- [test_mv_build_during_shutdown.debug.3](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/12835/testReport/junit/%28root%29/test_mv_fail_building/Tests___Unit_Tests___test_mv_build_during_shutdown_debug_3) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_mv_build_during_shutdown.debug.3)
#### Build Details:
- Duration: 3 hr 52 min
- Builder: spider1.cloudius-systems.com

denesb commented:
Test failures are caused by the new condition added by this PR being triggered:
```c++
    if (!_tasks.empty()) {
        on_fatal_internal_error(cmlog, format("{} tasks still exist after being stopped", _tasks.size()));
    }
```
This happens if the `stop_ongoing_compactions("shutdown")` above actually did stop a compaction. The `stop_ongoing_compactions()` will return as soon as the state of the stopped tasks is `done`. But at this point the tasks are still in `_tasks`, their removal races with the resume of `really_do_stop()`. If the race is lost, `really_do_stop()` will trip on the check, before the tasks are removed from `_tasks`.

denesb commented on compaction/compaction_manager.cc:
```diff
@@ -1134,16 +1134,23 @@ future<> compaction_manager::stop() {
+        cmlog.warn("Ignored error while stopping compaction tasks: {}", std::current_exception());
+    }
+    if (!_tasks.empty()) {
+        on_fatal_internal_error(cmlog, format("{} tasks still exist after being stopped", _tasks.size()));
```
This contradicts with the `catch (...)` above: we want to ignore errors during stop compaction, at the same time you assert that stop compaction finished. If there was an error, stop compaction will not have succeeded and it is not guaranteed that `_tasks` is empty.

denesb replied:
Ok, I see that `stop_tasks()` should call `stop()` on all tasks, even if some of them fail. So I guess this is fine. We just need to fix the race.

bhalevy replied:
maybe stop compactions should assert _tasks.empty() when done

bhalevy replied:
I think it should guarantee that.
The fact that a task completed with an error doesn't mean it wasn't properly stopped.
We should document that stop_ongoing_compactions and stop_tasks always succeed, and any error the may return, just represents the task completion status, not the status of stopping them.

bhalevy commented:
In v2 ([a338731](https://github.com/scylladb/scylladb/commit/a338731af20b8a67ee4d3c16461625f3886c4c2a)):
- rebased onto @denesb's branch, with those preliminary patches (that are a prerequisite for the internal error check at the last patch)
  - compaction/compaction_manager: make _tasks an intrusive list
  - compaction/compaction_manager: stop_tasks(): unlink stopped tasks
- Change Fixes tag to https://github.com/scylladb/scylladb/issues/21159 since https://github.com/scylladb/scylladb/issues/21298 is a dup of it
- and renamed subject line

scylladb-promoter commented:
#### :red_circle: CI State: FAILURE

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/12891/)
:white_check_mark: - [Docker Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci-docker/33/)
:white_check_mark: - [dtest with tablets](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-tablets/1464/)
:white_check_mark: - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/4535/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/12561/)
:x: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/12891/)
#### Failed Tests (3/36556):
- [scylla_gdb.run.release.1](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/12891/testReport/junit/%28root%29/non-boost%20tests/Tests___Unit_Tests___scylla_gdb_run_release_1) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+scylla_gdb.run.release.1)
- [scylla_gdb.run.release.2](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/12891/testReport/junit/%28root%29/non-boost%20tests/Tests___Unit_Tests___scylla_gdb_run_release_2) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+scylla_gdb.run.release.2)
- [scylla_gdb.run.release.3](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/12891/testReport/junit/%28root%29/non-boost%20tests/Tests___Unit_Tests___scylla_gdb_run_release_3) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+scylla_gdb.run.release.3)
#### Build Details:
- Duration: 3 hr 55 min
- Builder: spider7.cloudius-systems.com

denesb commented:
> * [scylla_gdb.run.release.1](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/12891/testReport/junit/%28root%29/non-boost%20tests/Tests___Unit_Tests___scylla_gdb_run_release_1) [🔍](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+scylla_gdb.run.release.1)
> 
> * [scylla_gdb.run.release.2](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/12891/testReport/junit/%28root%29/non-boost%20tests/Tests___Unit_Tests___scylla_gdb_run_release_2) [🔍](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+scylla_gdb.run.release.2)
> 
> * [scylla_gdb.run.release.3](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/12891/testReport/junit/%28root%29/non-boost%20tests/Tests___Unit_Tests___scylla_gdb_run_release_3) [🔍](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+scylla_gdb.run.release.3)

scylla-gdb.py needs to be adjusted for the changed type of `_tasks`. I updated my branch with the fix.

bhalevy commented:
> > * [scylla_gdb.run.release.1](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/12891/testReport/junit/%28root%29/non-boost%20tests/Tests___Unit_Tests___scylla_gdb_run_release_1) [🔍](https://github.com/scylladb/scylladb/issues?q=sort%3Aupdated-desc+is%3Aissue+is%3Aopen+scylla_gdb.run.release.1)
> > * [scylla_gdb.run.release.2](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/12891/testReport/junit/%28root%29/non-boost%20tests/Tests___Unit_Tests___scylla_gdb_run_release_2) [🔍](https://github.com/scylladb/scylladb/issues?q=sort%3Aupdated-desc+is%3Aissue+is%3Aopen+scylla_gdb.run.release.2)
> > * [scylla_gdb.run.release.3](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/12891/testReport/junit/%28root%29/non-boost%20tests/Tests___Unit_Tests___scylla_gdb_run_release_3) [🔍](https://github.com/scylladb/scylladb/issues?q=sort%3Aupdated-desc+is%3Aissue+is%3Aopen+scylla_gdb.run.release.3)
> 
> scylla-gdb.py needs to be adjusted for the changed type of `_tasks`. I updated my branch with the fix.

In [c9ed711](https://github.com/scylladb/scylladb/commit/c9ed7113fb7213d94d67907d7e8367c32f0b6a1b)

avikivity commented:
> 
> * Needs backport to 6.2 and 6.1, on top of [8cc9997](https://github.com/scylladb/scylladb/commit/8cc99973ebaf9d664c4c2a20b8ce3f1b2eb7100a)


This needs to explain why, not if. The labels already provide the if.

avikivity commented on compaction/compaction_manager.cc:
```diff
@@ -1138,16 +1138,23 @@ future<> compaction_manager::stop() {
-    co_await stop_ongoing_compactions("shutdown");
+    try {
+        co_await stop_ongoing_compactions("shutdown");
+    } catch (...) {
```
Do we expect stop_ongoing_compactions() to throw? IMO it should not.

raphaelsc replied:
Avi is right. I cannot remember of any usage of stop_ongoing_compactions() that requires exception to be propagated. Disabling compaction? no. Drain? no. Shutdown? no. For example, a failure in a task might cause problems in migration. Maybe nodetool stop compaction API? I don't remember if errors are propagated, but failed tasks are still stopped so returning success is still correct. Check stop_tasks().

bhalevy replied:
> Maybe nodetool stop compaction API? I don't remember if errors are propagated, but failed tasks are still stopped so returning success is still correct. Check stop_tasks().

Right.

Currently the tasks are indeed stopped, but if any of the failed, we return the error to the api caller.
I'm not sure if that's actually used by anyone (maybe dtests), but I wouldn't like to break the api.

We can add a an option to return the stopped tasks status which would be false by default and true only when called from the api layer.

bhalevy replied:
Since the nodetool documentation doesn't say anything about the returned status and the api is defined as `void`, it should be safe to just log any errors from tasks rather than returning them.

bhalevy commented:
In v3 ([6cce67b](https://github.com/scylladb/scylladb/commit/6cce67bec84f6cdebd09c450e1224d1e6aafd076)):
- Moved error handling down to `stop_tasks`
- Kept assert for empty tasks list after final stop_ongoing_compactions

bhalevy commented:
> > * Needs backport to 6.2 and 6.1, on top of [8cc9997](https://github.com/scylladb/scylladb/commit/8cc99973ebaf9d664c4c2a20b8ce3f1b2eb7100a)
> 
> This needs to explain why, not if. The labels already provide the if.

Added more context to the explanation.

scylladb-promoter commented:
#### :red_circle: CI State: FAILURE

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/12932/)
:white_check_mark: - [Docker Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci-docker/61/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/12599/)
:white_check_mark: - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/4574/)
:x: - [dtest with tablets](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-tablets/1501/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/12932/)
#### Build Details:
- Duration: 6 hr 20 min
- Builder: i-0c2ccbe4c5ae41afe (m5ad.12xlarge)

bhalevy commented:
> #### 🔴 CI State: FAILURE
> ✅ - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/12932/) ✅ - [Docker Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci-docker/61/) ✅ - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/12599/) ✅ - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/4574/) ❌ - [dtest with tablets](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-tablets/1501/) ✅ - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/12932/)

Unrelated issue: https://github.com/scylladb/scylladb/issues/21425
Respinned

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/12941/)
:white_check_mark: - [Docker Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci-docker/69/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/12609/)
:white_check_mark: - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/4584/)
:white_check_mark: - [dtest with tablets](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-tablets/1511/)
#### Build Details:
- Duration: 2 hr 21 min
- Builder: i-0728be3c0bae3bf54 (m5ad.8xlarge)

bhalevy commented:
@avikivity any more comments?
